### DISCUSSION TITLE
Changing photo labels does not propagate downstream

### DISCUSSION THREAD

**Discussion by youngdjn** [UTokyo-FieldPhenomics-Lab/EasyIDP#75] [General]
This is potentially a Python newbie issue -- apologies if so.

I am trying to change the labels of the photos (`Photo.label`) in my project, but this does not propagate to downstream places where I think the photo label is (should be?) used.

For example using [this project](https://ucdavis.box.com/s/fuxp9e9jtd7yjyxpbvnb1fqav106msdy), if I run:

```python
import easyidp as idp
from pathlib import Path
msproj = Path("/path/to/project/example-multifolder.psx")
ms = idp.Metashape(msproj, chunk_id=0)
ms.photos[57].label
```
I get  `'100MEDIA-DJI_0165'`

Then I attempt to rename them:
```python
for i in range(len(ms.photos)):
    ms.photos[i].label = "newlabel" + str(i) 
ms.photos[57].label
```
and I get `'newlabel57'`.

But then I run back projection
```python
img_dict_ms = roi.back2raw(ms)
img_dict_ms
```

And the labels in the returned dict are the original ones:

```
{'1': {'100MEDIA-DJI_0165': array([[4741.6161679 ,  684.46603374],
         [4735.11907978,  712.14368673],
         [4718.00310275,  728.6865289 ],
         [4722.27333877,  728.17479155],
         ...
```

And also when running `show_roi_on_img()` it only works when I reference the photos by their original labels, like:

```python
ms.show_roi_on_img(img_dict_ms, "1", "100MEDIA-DJI_0165")
```

and it doesn't work when I run:

```python
ms.show_roi_on_img(img_dict_ms, "1", "newlabel57")
```

Can you suggest a way to fully change the labels?

The reason for this request is that I am trying to give the photos more meaningful labels so that I can more easily link the photo labels to the original photo files.

**Comment by HowcanoeWang** [comment]
Actually, the `ms.photos` is a dictionary-like object, which allows you can assess the item by **BOTH** the `int` index (like a python list) and the `str` index (like as python dict), as a grammar sugar which python is not naturally supported:

```python
# list style
In [8]: ms.photos[0]    
Out[8]: <easyidp.reconstruct.Photo at 0x7fb119126130>

# dict style
In [9]: ms.photos['100MEDIA-DJI_0003']   
Out[9]: <easyidp.reconstruct.Photo at 0x7fb119126130>
```

The `ms.photos` object itself has the labels for indexing by `string`, you can check its keys by: 

```python
In [7]: ms.photos.keys()
Out[7]: dict_keys(['100MEDIA-DJI_0003', '100MEDIA-DJI_0004', '100MEDIA-DJI_0005', '100MEDIA-DJI_0006', ...
```

The easyidp uses the `ms.photos.keys()` as the output keys rather than the `ms.photos[i].labels`. (Apologies, I have never thought about the feature of changing image labels, it will be a debug disaster if the user rename two photo names with the same value by accident (and hard to detect by easyidp itself), the latter one will overwrite the previous one)

So in this case, if you really need to, you should change it carefully:

```python
for i in range(len(ms.photos)):
    new_key = "newlabel" + str(i) 
    old_key = ms.photos[i].label.copy()

    # change the photo.label
    # ms.photos[i].label = new_key

    # change the photos.key
    ms.photos.item_label[new_key] = ms.photos.item_label.pop(old_key)
```

Then the following ms object should return the expected labels for you.

```python
In [15]: ms.photos
Out[15]: 
<easyidp.Container> with 218 items
[0]	newlabel0
<easyidp.reconstruct.Photo object at 0x7fb119126130>
[1]	newlabel1
<easyidp.reconstruct.Photo object at 0x7fb119126f70>
...
[216]	newlabel216
<easyidp.reconstruct.Photo object at 0x7fb1087a16a0>
[217]	newlabel217
<easyidp.reconstruct.Photo object at 0x7fb1087a16d0>
```

---

Personally would prefer organizing the image folder / metashape camera group by meaningful name in the data preparation stage first, then using the produced name automatically, rather than changing them to be meaningful in the data analysis stage.

Hope this helps you.

---

Will support using photo.label as output index in the future release #76

**Reply by youngdjn** [reply]
Thank you, this all makes sense!
